File: benchmarks/pandas/bench_categorical_index_membership.py
"""
Benchmark: pandas.CategoricalIndex membership & ordering — category membership
("in" on categories), value containment ("in" on the index), and ordered
comparison (via categories.get_loc) over 100k lookups.
Outputs JSON: {"function": "categorical_index_membership", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 30

CATS = ["alpha", "beta", "gamma", "delta", "epsilon"]
labels = [CATS[i % len(CATS)] for i in range(SIZE)]
ci = pd.CategoricalIndex(labels, categories=CATS, ordered=True)
loc = ci.categories.get_loc


def run():
    has_count = 0
    contains_count = 0
    cmp_sum = 0
    for i in range(SIZE):
        label = CATS[i % len(CATS)]
        other = CATS[(i + 1) % len(CATS)]
        if label in ci.categories:
            has_count += 1
        if label in ci:
            contains_count += 1
        cmp_sum += loc(label) - loc(other)
    if has_count < 0 or contains_count < 0:
        raise RuntimeError("unreachable")
    return cmp_sum


for _ in range(WARMUP):
    run()

start = time.perf_counter()
for _ in range(ITERATIONS):
    run()
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "categorical_index_membership",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_categorical_index_membership.ts
/**
 * Benchmark: CategoricalIndex membership & ordering — hasCategory, contains,
 * and compareLabels (on an ordered index) over 100k lookups.
 * Outputs JSON: {"function": "categorical_index_membership", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { CategoricalIndex } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

const CATS = ["alpha", "beta", "gamma", "delta", "epsilon"];
const labels = Array.from({ length: SIZE }, (_, i) => CATS[i % CATS.length]);
const ci = CategoricalIndex.fromArray(labels, { categories: CATS, ordered: true });

function run(): void {
  let hasCount = 0;
  let containsCount = 0;
  let cmpSum = 0;
  for (let i = 0; i < SIZE; i++) {
    const label = CATS[i % CATS.length];
    if (ci.hasCategory(label)) hasCount++;
    if (ci.contains(label)) containsCount++;
    cmpSum += ci.compareLabels(label, CATS[(i + 1) % CATS.length]);
  }
  if (hasCount < 0 || containsCount < 0 || cmpSum === Number.NaN) {
    throw new Error("unreachable");
  }
}

for (let i = 0; i < WARMUP; i++) {
  run();
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  run();
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "categorical_index_membership",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
